src/transform/src/nonnull_requirements.rs

```diff
@@ -195,16 +195,80 @@ impl NonNullRequirements {
                 expected_group_size: _,
             } => {
                 let mut new_columns = HashSet::new();
-                for column in columns {
-                    // No obvious requirements on aggregate columns.
-                    // A "non-empty" requirement, I guess?
-                    if column < group_key.len() {
-                        group_key[column].non_null_requirements(&mut new_columns);
+                let (group_key_columns, aggr_columns): (Vec<usize>, Vec<usize>) =
+                    columns.iter().partition(|c| **c < group_key.len());
+                for column in group_key_columns {
+                    group_key[column].non_null_requirements(&mut new_columns);
+                }
+
+                if !aggr_columns.is_empty() {
+                    // For each aggregate, determine if it is both
+                    // 1. appropriate to consider as an aggregation that can push down non-null constraints, and
+                    // 2. which columns of the input being null would allow an input record to be discardable.
+                    let nonnull_info_per_aggregate = aggregates
+                        .iter()
+                        .enumerate()
+                        .map(|(pos, aggr)| {
+                            let should_push_down = aggr.func.propagates_nonnull_constraint()
+                                && aggr_columns.contains(&(group_key.len() + pos));
+                            let mut ignores_nulls_on_columns = HashSet::new();
+                            if let repr::Datum::Null = aggr.func.identity_datum() {
+                                aggr.expr
+                                    .non_null_requirements(&mut ignores_nulls_on_columns);
+                            }
+                            (should_push_down, ignores_nulls_on_columns)
+                        })
+                        .collect::<Vec<_>>();
+                    let mut pushable_nonnull_constraints: Option<HashSet<usize>> = None;
+                    for (should_push_down, ignores_nulls_on_columns) in
+                        nonnull_info_per_aggregate.iter()
+                    {
+                        // Each aggregate may either
```

Comment:
Hmm perhaps this is better explained with examples?
* Two aggs: `sum(#0 + #2), max(#0 + #1)`, non-null requirements on both aggs should be pushed down => implies `!isnull(#0)`. We don't want to push down a `!isnull(#2)` because deleting a row like `(1,1, null)` would make the max wrong.
* Two aggs: `sum(#0 + #2), max(#0 + #1)`, non-null requirements on only the max should be pushed down  => ???

Reply:
actually, the second case implies !isnull(#0) I wasn't covering that case. Thanks!